projects/plugins/jetpack/json-endpoints/class.wpcom-json-api-list-users-endpoint.php

```diff
@@ -164,87 +164,108 @@ public function callback( $path = '', $blog_id = 0 ) {
 			$query['capability'] = $args['capability'];
 		}
 
-		$user_query = new WP_User_Query( $query );
-
 		remove_filter( 'user_search_columns', array( $this, 'api_user_override_search_columns' ) );
 
+		$response        = array();
+		$users           = array();
+		$viewers         = array();
 		$is_wpcom        = defined( 'IS_WPCOM' ) && IS_WPCOM;
 		$include_viewers = (bool) isset( $args['include_viewers'] ) && $args['include_viewers'] && $is_wpcom;
-
-		$page    = ( (int) ( $args['offset'] / $args['number'] ) ) + 1;
-		$viewers = $include_viewers ? get_private_blog_users(
-			$blog_id,
-			array(
-				'page'     => $page,
-				'per_page' => $args['number'],
-			)
-		) : array();
-		$viewers = array_map( array( $this, 'get_author' ), $viewers );
-
-		// When include_viewers is true, search by username or email.
-		if ( $include_viewers && ! empty( $args['search'] ) ) {
-			$viewers = array_filter(
-				$viewers,
-				function ( $viewer ) use ( $args ) {
-					// Convert to WP_User so expected fields are available.
-					$wp_viewer = new WP_User( $viewer->ID );
-					// remove special database search characters from search term
-					$search_term = str_replace( '*', '', $args['search'] );
-					return ( str_contains( $wp_viewer->user_login, $search_term ) || str_contains( $wp_viewer->user_email, $search_term ) || str_contains( $wp_viewer->display_name, $search_term ) );
+		$is_multisite    = is_multisite();
+		$user_query      = new WP_User_Query( $query );
+
+		// Get users.
+		foreach ( $user_query->get_results() as $u ) {
+			$the_user = $this->get_author( $u, true );
+			if ( $the_user && ! is_wp_error( $the_user ) ) {
+				$userdata        = get_userdata( $u );
+				$the_user->roles = ! is_wp_error( $userdata ) ? array_values( $userdata->roles ) : array();
+				if ( $is_multisite ) {
+					$the_user->is_super_admin = user_can( $the_user->ID, 'manage_network' );
 				}
+				$users[] = $the_user;
+			}
+		}
+
+		// Get viewers.
+		if ( $include_viewers ) {
```

Comment:
I've refactored all this to test out grabbing, deduping and merging the user and viewer lists before we return arbitrary count values that are not accurate.

This would remove the need for frontend hacks like:

- https://github.com/Automattic/wp-calypso/pull/99448
- https://github.com/Automattic/wp-calypso/pull/50712/files#diff-d3b37aa7f029b8f881e02a9ea0d701be351599baeab1900624a99512954a6cceR43